docs/cudf/source/developer_guide/developer_guide.md

```diff
@@ -0,0 +1,129 @@
+# Developer Guide
+
+```{note}
+At present, this guide only covers the main cuDF library.
+In the future, it may be expanded to also cover dask_cudf, cudf_kafka, and custreamz.
+```
+
+The cuDF library is a GPU-accelerated, [Pandas-like](https://pandas.pydata.org/) DataFrame library.
+Under the hood, all of cuDF's functionality relies on the CUDA-accelerated `libcudf` C++ library.
+Thus, cuDF's internals are designed to efficiently and robustly map pandas APIs to `libcudf` functions.
+For more information about the `libcudf` library, a good starting point is the
+[developer guide](https://github.com/rapidsai/cudf/blob/main/cpp/docs/DEVELOPER_GUIDE.md).
+
+This document assumes familiarity with the
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md).
+The goal of this document is to provide more specific guidance for Python developers.
+It covers the structure of the Python code and discusses best practices.
+Additionally, it includes longer sections on more specific topics like testing and benchmarking.
+More specific information on these can be found in the pages below:
+
+```{toctree}
+:maxdepth: 1
+
+library_design
+```
+
+The rest of this document focuses on a higher-level overview of best practices in cuDF.
+
+## Directory structure and file naming
+
+cuDF generally presents the same importable modules and subpackages as pandas.
+All Cython code is contained in `python/cudf/cudf/_lib`.
+
+**Open question**: Should we start enforcing a stricter file/directory layout? Any suggestions?
```

Comment:
Coming in cold, probably

1. Layouts that can be enforced by linting rules
2. Layouts that avoid messy dependency chains (`/foo` only imports from `_lib`)

would probably be good start with. In my experience, soft conversions of structure without linting enforcement just tend to stray overtime

Reply:
I don't disagree with you there. This also came up in [the test discussion](https://github.com/rapidsai/cudf/pull/11199#discussion_r918283574), and it's a problem anywhere. Do you have any suggestions on what kind of layouts would be enforceable? I would love that if we could come up with something. CC @shwina

Reply:
Spitballing simple ones I can think of would be:

1. Mentioned in https://github.com/rapidsai/cudf/pull/11199#discussion_r918283152=, if the doc structure could match test structure (the rst parsing would be a little heavy though)
2. If `cudf/tests/**/` could mirror the structure of `cudf/**` directory without the test directory.

The weakness in both would be where to place integration tests that utilize multiple APIs together.